src/compiler/scala/tools/nsc/typechecker/Contexts.scala

```diff
@@ -1073,38 +1073,52 @@ trait Contexts { self: Analyzer =>
         found1
       }
 
-      def lookupInScope(scope: Scope) = {
-        val entries = scope lookupUnshadowedEntries name
-        val result = entries.filter(e => qualifies(e.sym)).toList
-        result
-      }
+      def lookupInScope(owner: Symbol, pre: Type, scope: Scope): Symbol = {
+        var e = scope.lookupEntry(name)
+        while (e != null && !qualifies(e.sym)) {
+          e = scope.lookupNextEntry(e)
+        }
+        if (e == null) {
+          NoSymbol
+        } else {
+          val e1 = e
+          val e1Sym = e.sym
+          var syms: mutable.ListBuffer[Symbol] = null
+          e = scope.lookupNextEntry(e)
+          while (e ne null) {
+            if (e.depth == e1.depth && e.sym != e1Sym && qualifies(e.sym)) {
+              if (syms eq null) {
+                syms = new mutable.ListBuffer[Symbol]
+                syms += e1Sym
+              }
+              syms += e.sym
+            }
+            e = scope.lookupNextEntry(e)
+          }
+          // we have a winner: record the symbol depth
+          symbolDepth = (cx.depth - cx.scope.nestingLevel) + e1.depth
```

Comment:
This does need to capture `symbolDepth` in a object ref, I can't see a good way to avoid that. It is only one per call to `lookupSymbol`.